Importing contacts

y-lohse opened the pull request:
This PR has all commits from #18 but a bunch more, handling contact imports.

Things to note:

- The `cozyUtils` file is almost the same than for the google connector. Not sure what to do about that.
- We update contacts on cozy, even if nothing has changed. We could detect this case and skip the updated, but I think we want to keep the `cozyMetadata.sync.lastSync` field accurate.

y-lohse commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(𝐌) [Connecteur Toutatice] 🎓- mapping & update contacts](https://trello.com/c/EXmqQhrS/1861-%F0%9D%90%8C-connecteur-toutatice-%F0%9F%8E%93-mapping-update-contacts)

Crash-- commented on package.json:
```diff
@@ -37,18 +38,25 @@
     "eslint-plugin-prettier": "3.0.1",
     "git-directory-deploy": "1.5.1",
     "husky": "1.3.1",
+    "json-loader": "^0.5.7",
```
^

Crash-- commented on src/CozyUtils.js:
```diff
@@ -0,0 +1,157 @@
+    log('error', 'Unable to initialize cozy client')
+    throw err
+  }
+}
```
On devrait clairement séparer : 
- le schema et l'init de CozyClient 
- la récupération du token et de l'URL 
- et le CozyUtils

Crash-- commented on src/synchronize.js:
```diff
@@ -0,0 +1,36 @@
+      remoteContact,
+      `cozyMetadata.sync.${contactAccountId}.id`
+    )
+    const cozyContact = await cozyUtils.findContact(contactAccountId, remoteId)
```
C'est prévu de laisser 1 requête pour chaque contact ? Je sais qu'on devrait pas en avoir des masses, mais même 100 requêtes, c'est pas forcément scalable

y-lohse replied:
C'était voulu parce que je pense que ca posera pas problème… mais a priori ca coute pas grand chose de récupérer tous les contacts cozy d'un seul coup. Je vais faire un essai.

Crash-- commented on src/synchronize.js:
```diff
@@ -0,0 +1,36 @@
+      cozyUtils.save(remoteContact)
+      result.contacts.created++
+    } else {
+      const merged = merge(cozyContact, remoteContact)
```
Je ne sais pas si on y a des possibilités de récupérer des `undefined` comme valeur pour des clés qui auront été supprimées côté toutatice ? Et aussi si es clés peuvent être supprimées côté toutatice ?

y-lohse replied:
A ma connaissance elles ne peuvent pas être supprimés, au mieux elles peuvent devenir `""`, que l'on veux utiliser comme valeur.
Et en cas de suppression complète, je pense que `merge` va prendre la valeur du cozy… mais je vais ajouter un test la dessus au cas ou.

Crash-- replied:
Yep, parce que si on se retrouve avec des `undefined` ou des clés supprimées côté toutatice, avec le merge de lodash on ne pourra pas les répercuter chez nous.

y-lohse replied:
Non mais je pense pas qu'on le veuille en fait ? Si une clé est supprimée côté toutatice, pour moi on laisse ce qu'il y a dans le cozy.

Crash-- replied:
Je sais pas, j'ai pas lu la spec et je ne sais pas comment ils vont gérer la suppression d'une information. Typiquement pour le contact ça devrait pas trop poser de problème, mais pour les groupes ?

cedricmessiant commented:
About `cozyUtils`, the reusable stuff should probably go in `cozy-konnector-libs` (See this issue : https://github.com/konnectors/libs/issues/415 )

cedricmessiant commented on src/index.js:
```diff
@@ -1,4 +1,6 @@
 const { BaseKonnector, log, errors } = require('cozy-konnector-libs')
+const fetch = require('node-fetch')
```
why do you need `node-fetch`?

y-lohse replied:
We need to get the list of contacts from a remote URL that is not a cozy server. `cozy-client-js` only allows to fetch cozy URLs, and I think it's the same for `cozy-client`.

cedricmessiant commented on src/transpileToCozy.js:
```diff
@@ -0,0 +1,53 @@
+}
+
+const transpileToCozy = (contact, contactsAccountsId) => {
+  const familyName = get(contact, 'firstname')
```
should be `lastname`

cedricmessiant commented on src/transpileToCozy.js:
```diff
@@ -0,0 +1,53 @@
+
+const transpileToCozy = (contact, contactsAccountsId) => {
+  const familyName = get(contact, 'firstname')
+  const givenName = get(contact, 'lastname')
```
and `firstname`

cedricmessiant commented:
Ok, I see. It may be nice that `lastSync` reflects the last sync even if there is no change but there is no easy way to do this update without changing `updatedAt` and other metadata. But we still have `lastSync` on the `io.cozy.contacts.accounts` and one could argue that the last sync of a contact could be the last time it has been actually synced. If we could avoid unnecessary mutations, it probably worth it.

cedricmessiant commented:
by the way, you should update the `io.cozy.contacts.accounts` fields after sync is done, right? (as we do in google konnector)

Crash-- commented:
> Ok, I see. It may be nice that lastSync reflects the last sync even if there is no change but there is no easy way to do this update without changing updatedAt and other metadata. But we still have lastSync on the io.cozy.contacts.accounts and one could argue that the last sync of a contact could be the last time it has been actually synced. If we could avoid unnecessary mutations, it probably worth it.

What is the behavior for the google konnector ?

y-lohse commented:
The google connector detects if there is a change to do or not (thanks to the `etag` and the `lastSync` fields), and if there is nothing to do, it leaves the contact alone.
Here we would have to check the fields coming from toutatice vs the ones on the cozy. It requires a bit of code change, but I think @cedricmessiant is right, the `lastSync` on the contact should reflect the last time something *changed*.

> by the way, you should update the io.cozy.contacts.accounts fields after sync is done, right? (as we do in google konnector)

We don't use the `io.cozy.contacts.accounts.lastSync` here… but yes, it would make sense to do it anyway 👍

Crash-- commented:
If we update the `lastSync` with Google only when there is a change, we should do the same here.

Since we all agreed on that, maybe we edit our doctype documentation?

y-lohse commented:
Had to force push, but the changes since the last review round start with https://github.com/konnectors/toutatice/pull/23/commits/d5e7e1d85a100c84d0bc651ee6e2363c6c737b42

cedricmessiant commented on src/helpers/transpileToCozy.spec.js:
```diff
@@ -124,4 +124,32 @@ describe('Transpile to cozy', () => {
     expect(result.cozy).toEqual([])
   })
+
+  it('should handle mising fields', () => {
```
`mising` :arrow_right: `missing`
